### PR TITLE
Picolibc build with llvm-toolchain, run tests.

### PR DESCRIPTION
**Pull request by pwprzybyla**
Picolibc tests build, executed with shell script

Picolibc tests executed with test parameter

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -532,6 +488,88 @@ while(library_variants_copy)
...
+        set(picolibc_tests printf_scanf
+                           printf-tests
+                           rand
+                           regex
+                           ungetc
+                           fenv
+                           long_double
+                           math_errhandling
+                           malloc
+                           tls
```
Comment (mplatings) [inline CMakeLists.txt:545]: I guess tls needs commenting out since you've disabled building it

**Review by mplatings — CHANGES_REQUESTED** [review]
This is great to see!
I've commented with various nits but over all it looks very promising.

File: CMakeLists.txt [hunk 2]
```diff
@@ -110,7 +110,6 @@
 # CONFIGURE_HANDLED_BY_BUILD was introduced in CMake 3.20 and it
 # greatly speeds up incremental builds.
 cmake_minimum_required(VERSION 3.20)
-
```
Comment (mplatings) [inline CMakeLists.txt:113]: Avoid unnecessary changes like this

File: CMakeLists.txt [hunk 3]
```diff
@@ -382,27 +381,26 @@ endfunction()
...
-    armv8.1m.main   hard_fp             "-mfloat-abi=hard -march=armv8.1m.main+fp"
-    armv8.1m.main   hard_nofp_mve       "-mfloat-abi=hard -march=armv8.1m.main+nofp+mve"
-    armv8.1m.main   soft_nofp_nomve     "-mfloat-abi=soft -march=armv8.1m.main+nofp+nomve"
+    armv6m          soft_nofp           "-mfloat-abi=soft -march=armv6m"                        "${picolibc_SOURCE_DIR}/scripts/run-thumbv6m"
+    armv7em         hard_fpv4_sp_d16    "-mfloat-abi=hard -march=armv7em -mfpu=fpv4-sp-d16"     "${picolibc_SOURCE_DIR}/scripts/run-thumbv7m"
+    armv7em         hard_fpv5_d16       "-mfloat-abi=hard -march=armv7em -mfpu=fpv5-d16"        "${picolibc_SOURCE_DIR}/scripts/run-thumbv7m"
+    armv7em         soft_nofp           "-mfloat-abi=soft -march=armv7em -mfpu=none"            "${picolibc_SOURCE_DIR}/scripts/run-thumbv7m"
+    armv7m          soft_nofp           "-mfloat-abi=soft -march=armv7m+nofp"                   "${picolibc_SOURCE_DIR}/scripts/run-thumbv7m"
+    armv8m.main     hard_fp             "-mfloat-abi=hard -march=armv8m.main+fp"                "${picolibc_SOURCE_DIR}/scripts/run-arm"
```
Comment (mplatings) [inline CMakeLists.txt:392]: Does run-arm need extra arguments to select the sub-architecture?

File: CMakeLists.txt [hunk 4]
```diff
@@ -382,27 +381,26 @@ endfunction()
...
+    armv8.1m.main   soft_nofp_nomve     "-mfloat-abi=soft -march=armv8.1m.main+nofp+nomve"      "${picolibc_SOURCE_DIR}/scripts/run-arm"
 )
 
-
+add_custom_target(check-picolibc)
 set(library_variants_copy "${library_variants}")
 # Destructively iterate over a copy of the tuples defining the variants.
 while(library_variants_copy)
-    list(POP_FRONT library_variants_copy parent_dir_name libc_type target_arch variant flags)
-
```
Comment (mplatings) [inline CMakeLists.txt:405]: Avoid unnecessary changes like this

File: CMakeLists.txt [hunk 5]
```diff
@@ -432,7 +430,6 @@ while(library_variants_copy)
     if(CMAKE_INSTALL_MESSAGE STREQUAL NEVER)
         set(MESON_INSTALL_QUIET "--quiet")
     endif()
-
```
Comment (mplatings) [inline CMakeLists.txt:435]: Avoid unnecessary changes like this

File: CMakeLists.txt [hunk 6]
```diff
@@ -442,7 +439,7 @@ while(library_variants_copy)
             DEPENDS clang lld llvm-ar llvm-config llvm-nm llvm-ranlib llvm-strip
             CONFIGURE_COMMAND ${MESON_EXECUTABLE} -Dincludedir=include -Dlibdir=lib --prefix <INSTALL_DIR> --cross-file <BINARY_DIR>/meson-cross-build.txt ${picolibc_SOURCE_DIR}
             BUILD_COMMAND ninja
-            INSTALL_COMMAND ${MESON_EXECUTABLE} install ${MESON_INSTALL_QUIET}
+            INSTALL_COMMAND meson install ${MESON_INSTALL_QUIET}
```
Comment (mplatings) [inline CMakeLists.txt:442]: Change this back. You can specify MESON_EXECUTABLE with a -D argument if cmake can't find meson automatically.

File: CMakeLists.txt [hunk 7]
```diff
@@ -457,6 +454,7 @@ while(library_variants_copy)
         set(meson_c_args "${common_flags}")
         string(REPLACE " " "', '" meson_c_args "${meson_c_args}")
         set(meson_c_args "'${llvm_bin}/clang', '${meson_c_args}'")
+
```
Comment (mplatings) [inline CMakeLists.txt:457]: Avoid unnecessary changes like this

File: CMakeLists.txt [hunk 8]
```diff
@@ -532,6 +530,82 @@ while(library_variants_copy)
             ${INSTALL_DIR}/lib/${target} ${install_directory}/lib
     )
 
+    if(libc_type STREQUAL "picolibc")
+        ExternalProject_Add(
```
Comment (mplatings) [inline CMakeLists.txt:534]: It seems you're adding a second picolibc build with tests enabled. I don't like it, particularly as building picolibc takes a non-trivial amount of time.

The picolibc tests seem very simple. If there isn't a clean way to build the tests with meson then I'd be OK with building them with CMake, even if that means duplicating a lot of https://github.com/picolibc/picolibc/blob/main/test/CMakeLists.txt

File: CMakeLists.txt [hunk 9]
```diff
@@ -674,7 +747,7 @@ else()
     # Also run install-distribution-stripped to install the LLVM
     # binaries.
     add_dependencies(
-        install-llvm-toolchain
+        install-llvm-toolchain
```
Comment (mplatings) [inline CMakeLists.txt:750]: Avoid unnecessary whitespace changes

File: patches/picolibc-HEAD.patch [hunk 10]
```diff
@@ -15,3 +39,65 @@ index 4bb20f3af..cf9e564d1 100644
...
++	/*PROVIDE( __tbss_offset = ADDR(.tbss) - ADDR(.tdata) );*/
+ 	PROVIDE( __tbss_size = SIZEOF(.tbss) );
+ 	PROVIDE( __tls_size = __tls_end - __tls_base );
+ 	PROVIDE( __tls_align = MAX(ALIGNOF(.tdata), ALIGNOF(.tbss)) );
+@@ -208,7 +208,7 @@ SECTIONS
+ 	 * so we create a special segment here just to make room
+ 	 */
+ 	.tbss_space (NOLOAD) : {
+-		. = ADDR(.tbss);
++		/*. = ADDR(.tbss);*/
```
Comment (mplatings) [inline patches/picolibc-HEAD.patch:87]: If you rebase your patch on the latest version of LLVM-embedded-toolchain-for-Arm I think you'll find this is unnecessary.

**Review by mplatings — CHANGES_REQUESTED** [review]
Please mark my comments as resolved in the GitHub UI once you think they're addressed.

File: CMakeLists.txt [hunk 11]
```diff
@@ -165,7 +165,7 @@ endif()
 # If lld is available then use it by default.
 find_program(LLD_EXECUTABLE lld)
 if(LLD_EXECUTABLE)
-    set(LLVM_ENABLE_LLD ON CACHE BOOL "")
+    set(LLVM_ENABLE_LLD OFF CACHE BOOL "")
```
[the remark this answers is not in the document]
Comment (pwprzybyla) [reply CMakeLists.txt]: Ok.

File: patches/picolibc_tests.patch [hunk 12]
```diff
@@ -0,0 +1,67 @@
+From d2042093df556d3543bcc36e106c3e67e6488338 Mon Sep 17 00:00:00 2001
```
[the remark this answers is not in the document]
Comment (pwprzybyla) [reply patches/picolibc_tests.patch]: Not needed any more.

File: CMakeLists.txt [hunk 13]
```diff
@@ -731,7 +768,6 @@ add_dependencies(
 
 # Run tests on the built toolchain.
 add_custom_target(check-llvm-toolchain)
-
```
[the remark this answers is not in the document]
Comment (pwprzybyla) [reply CMakeLists.txt:734]: Done.

File: CMakeLists.txt (hunk 3 above)
Comment (pwprzybyla) [reply CMakeLists.txt:392]: It's pretty complex script that detects and then sets variety of qemu parameters. That might be the reason it works for amrv7em_soft_nofp and for others it sets wrong parameters.

File: CMakeLists.txt (hunk 1 above)
Comment (pwprzybyla) [reply CMakeLists.txt]: true

File: CMakeLists.txt (hunk 4 above)
Comment (pwprzybyla) [reply CMakeLists.txt:405]: Ok.

File: CMakeLists.txt (hunk 6 above)
Comment (pwprzybyla) [reply CMakeLists.txt:442]: Indeed merge issue.

File: CMakeLists.txt (hunk 8 above)
Comment (pwprzybyla) [reply CMakeLists.txt]: Done.

File: patches/picolibc-HEAD.patch (hunk 10 above)
Comment (pwprzybyla) [reply patches/picolibc-HEAD.patch]: Thanks.

File: CMakeLists.txt [hunk 14]
```diff
@@ -532,6 +488,88 @@ while(library_variants_copy)
...
+    if(CMAKE_INSTALL_MESSAGE STREQUAL NEVER)
+        set(MESON_INSTALL_QUIET "--quiet")
+    endif()
+    if(libc_type STREQUAL "picolibc")
+        ExternalProject_Add(
+            picolibc_${variant}
+            SOURCE_DIR ${picolibc_SOURCE_DIR}
+            INSTALL_DIR ${install_directory}
+            PREFIX picolibc/${variant}
+            DEPENDS clang lld llvm-ar llvm-config llvm-nm llvm-ranlib llvm-strip compiler_rt_${variant}
```
[the remark this answers is not in the document]
Comment (pwprzybyla) [reply CMakeLists.txt]: Ok.

File: CMakeLists.txt [hunk 15]
```diff
@@ -532,6 +530,57 @@ while(library_variants_copy)
...
+                        time-tests
+                        test-strtod
+                        test-strchr
+                        test-memset
+                        test-put
+                        test-efcvt
+                        malloc_stress
+                        posix-io)
+    foreach(test IN LISTS picolibc_tests)
+        add_custom_target(
```
Comment (mplatings) [inline CMakeLists.txt:563]: Use add_custom_command here. add_custom_target is always considered out of date: https://cmake.org/cmake/help/latest/command/add_custom_target.html

File: CMakeLists.txt [hunk 16]
```diff
@@ -532,6 +530,57 @@ while(library_variants_copy)
...
+                ${install_directory}/lib/libm.a ${install_directory}/lib/libclang_rt.builtins.a
+                ${install_directory}/lib/libsemihost.a
+                ${picolibc_SOURCE_DIR}/test/${test}.c -o picolibc_test_${variant}_${test}.out
+            COMMENT "Building: ${test} for ${variant}"
+        )
+        add_custom_target(
+                check-picolibc_${variant}_${test}
+                ../samples/picolibc_test_wrapper.sh ${picolibc_test_command} picolibc_test_${variant}_${test}.out
+                DEPENDS picolibc_${variant} build-check-picolibc_${variant}_${test}
+                USES_TERMINAL
```
Comment (mplatings) [inline CMakeLists.txt:577]: `USES_TERMINAL` prevents tests running in parallel. The CMake docs say "With the [Ninja](https://cmake.org/cmake/help/latest/generator/Ninja.html#generator:Ninja) generator, this places the command in the console [pool](https://cmake.org/cmake/help/latest/prop_gbl/JOB_POOLS.html#prop_gbl:JOB_POOLS)" - I infer that only one job in the console pool can run at a time.

File: CMakeLists.txt [hunk 17]
```diff
@@ -532,6 +530,57 @@ while(library_variants_copy)
             ${INSTALL_DIR}/lib/${target} ${install_directory}/lib
     )
 
+    #picolibc tests
+    string(REPLACE " " ";" clang_compile_flags ${runtimes_flags})
+    add_custom_target(check-picolibc_${variant})
+    set(picolibc_test_ld ${picolibc_SOURCE_DIR}/scripts/test-arm.ld)
+    if(${variant} EQUAL "aarch64")
+        set(picolibc_test_ld ${picolibc_SOURCE_DIR}/scripts/test-aarch64.ld)
```
Comment (mplatings) [inline CMakeLists.txt:538]: Since AArch64 tests don't seem to be working you could try using `samples/ldscripts/raspi3b.ld`. See #169

**Review by mplatings — CHANGES_REQUESTED** [review]

**Review by pwprzybyla — COMMENTED** [review]
Agree on all.

**Review by mplatings — CHANGES_REQUESTED** [review]
This looks really good. My comments are just cosmetic at this point

File: patches/picolibc.patch [hunk 18]
```diff
@@ -32,3 +32,56 @@ index c757f7665..e262e1c21 100644
  
  	.bss (NOLOAD) : {
  		*(.sbss*)
+diff --git a/picolibc.ld.in b/picolibc.ld.in
```
Comment (mplatings) [inline patches/picolibc.patch:35]: This is a duplicate of line 20 of the patch file and it prevents the patch from applying cleanly. While running `git apply` I get `error: patch failed: picolibc.ld.in:207`

File: patches/picolibc.patch [hunk 19]
```diff
@@ -32,3 +32,56 @@ index c757f7665..e262e1c21 100644
...
+diff --git a/test/rounding-mode-sub.c b/test/rounding-mode-sub.c
+index 58372d9ef..8757760da 100644
+--- a/test/rounding-mode-sub.c
++++ b/test/rounding-mode-sub.c
+@@ -38,7 +38,7 @@
+ #include <fenv.h>
+ 
+ double
+-div(double a, double b)
++div__(double a, double b)
```
Comment (mplatings) [inline patches/picolibc.patch:62]: Why?

File: samples/ldscripts/aarch64.ld [hunk 20]
```diff
@@ -0,0 +1,15 @@
+/*
+Raspberry Pi does not have flash, but to accommodate for the picolibc.ld
```
Comment (mplatings) [inline samples/ldscripts/aarch64.ld:2]: If this file is intended to be the same as raspi3b.ld then just INCLUDE that. If it's intended to vary independently then update the comment.

File: CMakeLists.txt [hunk 21]
```diff
@@ -495,6 +495,104 @@ while(library_variants_copy)
...
+            )
+            add_custom_target(
+                    check-picolibc_${variant}_${test}
+                    ${qemu_command} ${qemu_params} -semihosting -nographic -kernel check-picolibc_${variant}_build_${test}.out
+                    DEPENDS picolibc_${variant} check-picolibc_${variant}_build_${test}.out
+                    COMMENT "TEST: ${variant} ${test}"
+            )
+            add_dependencies(check-picolibc_${variant} check-picolibc_${variant}_${test})
+        endforeach()
+        add_custom_command(
```
Comment (mplatings) [inline CMakeLists.txt:563]: A lot of duplication here. Use a function like https://github.com/picolibc/picolibc/blob/main/test/CMakeLists.txt#L67

**Review by mplatings — CHANGES_REQUESTED** [review]

File: patches/picolibc.patch (hunk 19 above)
Comment (pwprzybyla) [reply patches/picolibc.patch:62]: Without this patch test cannot be compiled due to duplicate symbol:
check-picolibc_armv6m_soft_nofp_build_rounding-mode.out 
cd /home/pioprz01/work_repo/build && /home/pioprz01/work_repo/build/llvm/bin/clang --config /home/pioprz01/work_repo/build/llvm/bin/armv6m_soft_nofp_semihost.cfg -T ../samples/ldscripts/armv6m_soft_nofp.ld -nostdlib -lc -lg -lm -lclang_rt.builtins -lsemihost -Oz /data_sda/pioprz01/LLVM-embedded-toolchain-for-Arm_fork/LLVM-embedded-toolchain-for-Arm/repos/picolibc/test/rounding-mode.c /data_sda/pioprz01/LLVM-embedded-toolchain-for-Arm_fork/LLVM-embedded-toolchain-for-Arm/repos/picolibc/test/rounding-mode-sub.c -o check-picolibc_armv6m_soft_nofp_build_rounding-mode.out
ld.lld: error: duplicate symbol: div
>>> defined at div.c:82 (/home/pioprz01/work_repo/build/picolibc/armv6m_soft_nofp/src/picolibc_armv6m_soft_nofp-build/../../../../../repos/picolibc/newlib/libc/stdlib/div.c:82)
>>>            libc_stdlib_div.c.o:(div) in archive /home/pioprz01/work_repo/build/llvm/bin/../lib/clang-runtimes/arm-none-eabi/armv6m_soft_nofp/lib/libc.a
>>> defined at rounding-mode-sub.c
>>>            /tmp/rounding-mode-sub-099cc0.o:(.text+0x1)
clang-16: error: ld.lld command failed with exit code 1 (use -v to see invocation)

File: samples/ldscripts/aarch64.ld (hunk 20 above)
Comment (pwprzybyla) [reply samples/ldscripts/aarch64.ld:2]: Idea behind this approach is to have linker script for each variant. Just like with .cfg files. Indeed currently it is a copy of raspi3b.ld but may change in the future. More to say with naming in format ${variant}.ld there is no need to have mapping between variant and script file.

File: patches/picolibc.patch (hunk 18 above)
Comment (pwprzybyla) [reply patches/picolibc.patch:35]: fixed.

File: patches/picolibc.patch (hunk 19 above)
Comment (mplatings) [reply patches/picolibc.patch:62]: Thanks for explaining. Yes, there's a standard C `div` function. I think this is worth a PR in picolibc. For now I'd prefer to leave this test disabled, and not patch picolibc.

File: patches/picolibc.patch (hunk 18 above)
Comment (mplatings) [reply patches/picolibc.patch:35]: Please ensure that the code has actually been changed in the pull request before commenting that something has been fixed. Maybe in your local copy you've fixed it, but as far as everyone else is concerned the error remains and your comment is incorrect.

File: CMakeLists.txt [hunk 22]
```diff
@@ -495,6 +495,76 @@ while(library_variants_copy)
...
+                    OUTPUT check-picolibc_${variant}_build_${test}.out
+                    COMMAND ${llvm_bin}/clang --config ${llvm_bin}/${variant}_semihost.cfg
+                    -T ../samples/ldscripts/${variant}.ld
+                    -nostdlib -lc -lg -lm -lclang_rt.builtins -lsemihost -Oz
+                    ${picolibc_SOURCE_DIR}/test/${test}.c ${ARGN} -o check-picolibc_${variant}_build_${test}.out
+                    COMMENT "Building: ${test} for ${variant}"
+                    DEPENDS ${picolibc_SOURCE_DIR}/test/${test}.c
+            )
+            add_custom_target(
+                    check-picolibc_${variant}_${test}
```
Comment (mplatings) [inline CMakeLists.txt:553]: Change the name to `check-picolibc_${variant}-${test}`

File: CMakeLists.txt [hunk 23]
```diff
@@ -495,6 +495,76 @@ while(library_variants_copy)
             -DLIBCXX_ENABLE_WIDE_CHARACTERS=OFF
             -DLIBUNWIND_ENABLE_THREADS=OFF
         )
+
+        #picolibc tests
+        string(REPLACE " " ";" qemu_params "${qemu_params}")
+        add_custom_target(check-picolibc_${variant})
+        set(qemu_command "qemu-system-arm")
```
Comment (mplatings) [inline CMakeLists.txt:502]: This can be changed to `set(qemu_command "qemu-system-${cpu_family}")` and the following `if` statement can be removed.

File: CMakeLists.txt [hunk 24]
```diff
@@ -495,6 +495,76 @@ while(library_variants_copy)
...
+        if("${variant}" STREQUAL "aarch64")
+            set(qemu_command "qemu-system-aarch64")
+        endif()
+        set(picolibc_tests  abort
```
Comment (mplatings) [inline CMakeLists.txt:506]: Moving `abort` to the next line will improve readability slightly.

File: CMakeLists.txt [hunk 25]
```diff
@@ -495,6 +495,76 @@ while(library_variants_copy)
...
+                    test-mktemp
+                    test-put
+                    test-strchr
+                    test-strtod
+                    timegm
+                    time-sprintf
+                    time-tests
+                    tls
+                    ungetc
+                    )
```
Comment (mplatings) [inline CMakeLists.txt:540]: Unindent, for consistency with the style of other code.

File: CMakeLists.txt [hunk 26]
```diff
@@ -495,6 +495,76 @@ while(library_variants_copy)
...
+                    DEPENDS picolibc_${variant} check-picolibc_${variant}_build_${test}.out
+                    COMMENT "TEST: ${variant} ${test}"
+            )
+            add_dependencies(check-picolibc_${variant} check-picolibc_${variant}_${test})
+        endfunction()
+
+        foreach(test IN LISTS picolibc_tests)
+            picolibc_test(${test})
+        endforeach()
+        #Test disabled due to duplicated symbol name div in test file and in picolibc
```
Comment (mplatings) [inline CMakeLists.txt:564]: Space after `#`

File: CMakeLists.txt [hunk 27]
```diff
@@ -495,6 +495,76 @@ while(library_variants_copy)
             -DLIBCXX_ENABLE_WIDE_CHARACTERS=OFF
             -DLIBUNWIND_ENABLE_THREADS=OFF
         )
+
+        #picolibc tests
```
Comment (mplatings) [inline CMakeLists.txt:499]: Space after `#`

File: CMakeLists.txt [hunk 28]
```diff
@@ -495,6 +495,76 @@ while(library_variants_copy)
...
+                    -T ../samples/ldscripts/${variant}.ld
+                    -nostdlib -lc -lg -lm -lclang_rt.builtins -lsemihost -Oz
+                    ${picolibc_SOURCE_DIR}/test/${test}.c ${ARGN} -o check-picolibc_${variant}_build_${test}.out
+                    COMMENT "Building: ${test} for ${variant}"
+                    DEPENDS ${picolibc_SOURCE_DIR}/test/${test}.c
+            )
+            add_custom_target(
+                    check-picolibc_${variant}_${test}
+                    ${qemu_command} ${qemu_params} -semihosting -nographic -kernel check-picolibc_${variant}_build_${test}.out
+                    DEPENDS picolibc_${variant} check-picolibc_${variant}_build_${test}.out
```
Comment (mplatings) [inline CMakeLists.txt:555]: Remove the dependency on `picolibc_${variant}`

**Review by mplatings — CHANGES_REQUESTED** [review]
Looking really good. Just some nits and I think we'll be done.

**Review by pwprzybyla — COMMENTED** [review]
Done.

File: CMakeLists.txt [hunk 29]
```diff
@@ -495,6 +495,76 @@ while(library_variants_copy)
...
+            test-strchr
+            test-strtod
+            timegm
+            time-sprintf
+            time-tests
+            tls
+            ungetc
+        )
+
+        function (picolibc_test test)
```
Comment (mplatings) [inline CMakeLists.txt:540]: The linter doesn't like the space after `function`

Comment (pwprzybyla) [reply CMakeLists.txt]: Fixed.

File: CMakeLists.txt [hunk 30]
```diff
@@ -495,6 +495,76 @@ while(library_variants_copy)
...
+            # -Oz parameter is added just like picolibc is doing, otherwise there is a build issue
+            # ld.lld: error: undefined symbol: __issignalingl  while building math_errhandling test.
+            add_custom_command(
+                    OUTPUT check-picolibc_${variant}_build_${test}.out
+                    COMMAND ${llvm_bin}/clang --config ${llvm_bin}/${variant}_semihost.cfg
+                    -T ../samples/ldscripts/${variant}.ld
+                    -nostdlib -lc -lg -lm -lclang_rt.builtins -lsemihost -Oz
+                    ${picolibc_SOURCE_DIR}/test/${test}.c ${ARGN} -o check-picolibc_${variant}_build_${test}.out
+                    COMMENT "Building: ${test} for ${variant}"
+                    DEPENDS ${picolibc_SOURCE_DIR}/test/${test}.c ${libc_target}
```
Comment (mplatings) [inline CMakeLists.txt:550]: This is missing a dependency on `compiler_rt_${variant}`.

File: samples/ldscripts/armv7em_hard_fpv4_sp_d16.ld [hunk 31]
```diff
@@ -0,0 +1,7 @@
+__flash =      0x00000000;
+__flash_size = 0x00400000;
+__ram =        0x20000000;
+__ram_size   = 0x00200000;
+__stack_size = 4k;
+
+INCLUDE picolibc.ld
```
Comment (mplatings) [inline samples/ldscripts/armv7em_hard_fpv4_sp_d16.ld:7]: Add a terminating newline to this and the other .ld files

**Review by pwprzybyla — COMMENTED** [review]
PR changes applied.
Biggest subset of tests left either by disabling the failing tests or variant that fails most of the tests.
Additionally tests from different variants are stored in unique folders so they can run in parallel.

File: CMakeLists.txt [hunk 32]
```diff
@@ -495,6 +495,84 @@ while(library_variants_copy)
...
+            time-tests
+            #tls
+            ungetc
+        )
+
+        function(picolibc_test test)
+            # -Oz parameter is added just like picolibc is doing, otherwise there is a build issue
+            # ld.lld: error: undefined symbol: __issignalingl  while building math_errhandling test.
+            add_custom_command(
+                    OUTPUT check-picolibc_${variant}_build_${test}.out
```
Comment (mplatings) [inline CMakeLists.txt:545]: I think the output path is now incorrect

Comment (pwprzybyla) [reply CMakeLists.txt:545]: Actually it is just a string that has to match one used in DEPENDS that is few lines below. I did test with completely dummy name and it worked. So functionally it seems to be correct. Let me know if you want me to change it.

Comment (mplatings) [reply CMakeLists.txt:545]: According to the docs, OUTPUT should "Specify the output files the command is expected to produce".
CMake has a habit of making things work in most circumstances, and not quite working in others, and I suspect this is one of those. For example, I wouldn't be surprised to find that the target gets rebuilt every time because an absent file is treated the same as it being out of date. So yes, change it please.

Comment (pwprzybyla) [reply CMakeLists.txt]: Ok.

**Review by mplatings — APPROVED** [review]
